wave_lang/kernel/wave/decompose_topk_ops.py

```diff
@@ -0,0 +1,615 @@
+            get_custom(init_val), ["cond", "if_true", "if_false"]
+        )
+        resolve_broadcasting_for_op(
+            get_custom(init_idx), ["cond", "if_true", "if_false"]
+        )
+    return init_val, init_idx
+
+
+def emit_variable_topk_reduction(
+    binary_fn: Callable,
+    src_val: fx.Node,
+    src_idx: fx.Node,
+    graph: fx.Graph,
+    local_reduction_size: int,
+    location: Optional[CapturedLocation],
+) -> tuple[fx.Node, fx.Node]:
+    """
+    Does topk reduction over singular value and index variables.
+    """
+    init_val = get_graph_node(Extract(src_val, [0]), graph, location)
+    init_idx = get_graph_node(Extract(src_idx, [0]), graph, location)
+    get_custom(init_val).index = get_custom(src_val).index
+    get_custom(init_idx).index = get_custom(src_idx).index
+    # From here on out we are dealing with scalar values
+    init_val.type.symbolic_shape = ()
+    init_idx.type.symbolic_shape = ()
+
+    for i in range(1, local_reduction_size):
+        cur_val = get_graph_node(Extract(src_val, [i]), graph, location)
+        cur_idx = get_graph_node(Extract(src_idx, [i]), graph, location)
+        cur_val.type.symbolic_shape = ()
+        cur_idx.type.symbolic_shape = ()
+
+        cmp_result = get_graph_node(Gt(cur_val, init_val), graph, location)
+        init_val = get_graph_node(
+            SelectOp(cmp_result, cur_val, init_val), graph, location
+        )
+        init_idx = get_graph_node(
+            SelectOp(cmp_result, cur_idx, init_idx), graph, location
+        )
+        get_custom(init_val).index = get_custom(cur_val).index
+        get_custom(init_idx).index = get_custom(cur_idx).index
+
+    return init_val, init_idx
+
+
+def emit_local_topk_reduction(
+    binary_fn: Callable,
+    reduction_src: list[fx.Node],
+    reduction_indices: list[fx.Node],
+    graph: fx.Graph,
+    local_reduction_size,
+    location: Optional[CapturedLocation],
+) -> tuple[fx.Node, fx.Node]:
+    """
+    Does topk reduction over all elements carried along by TopkOp at local
+    thread level. This returns both the top value and its index.
+    """
+    src_val, src_idx = emit_sources_topk_reduction(
+        binary_fn, reduction_src, reduction_indices, graph, location
+    )
+
+    local_val, local_idx = emit_variable_topk_reduction(
+        binary_fn, src_val, src_idx, graph, local_reduction_size, location
+    )
+    return local_val, local_idx
+
+
+def emit_intrawave_topk_reduction(
+    binary_fn: Callable,
+    src_val: fx.Node,
+    src_idx: fx.Node,
+    graph: fx.Graph,
+    subgroup_size: int,
+    cluster_size: int,
+    cluster_stride: int,
+    location: Optional[CapturedLocation],
+) -> tuple[fx.Node, fx.Node]:
+    """
+    Reduce data across threads in a wave by doing butterfly shuffle for topk.
+    """
+    init_val = src_val
+    init_idx = src_idx
+    num_steps = int(math.log2(float(cluster_size)))
+
+    for _ in range(num_steps):
+        # Shuffle both values and indices
```

Comment:
No action required, but maybe you can add a TODO - if the values are f16 and the indices are <= f16, we could pack them together and to a single shuffle with both and this would reduce one extra shuffle required.